### PR TITLE
[FEATURE][needs-doc] Vertex tool can work on the current layer only

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

This PR adds an option to use the `Vertex Tool` on the active layer only instead of on all layers.

A dropdown menu allows to select the corresponding tool:

Unit tests are coming soon.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by m-kuhn** [comment]
Is this an application level or project level option?
Would it make sense to add a keyboard modifier to toggle this behaviour on the fly?

**Review by 3nids — COMMENTED** [review]
I would avoid adding a global option which makes a qgis map tool behave differently. 
I would prefer a key modifier on the tool.
Or if it's really requested a sub icon/action for the map tool for the new behavior (rather than adding a new icon to the snapping toolbar dedicated to a single map tool)

File: src/core/qgssnappingconfig.h [hunk 1]
```diff
@@ -172,6 +172,10 @@ class CORE_EXPORT QgsSnappingConfig
     //! set if the snapping on intersection is enabled
     void setIntersectionSnapping( bool enabled );
 
+    bool editVerticesOnAllLayers() const;
+
+    void setEditVerticesOnAllLayers( bool enabled );
+
```
Comment (wonder-sk) [inline src/core/qgssnappingconfig.h]: This setting should not be a part of QgsSnappingConfig as it is not related to snapping at all... It is an option specific to one particular tool...

File: src/app/qgssnappingwidget.cpp [hunk 2]
```diff
@@ -199,6 +199,14 @@ QgsSnappingWidget::QgsSnappingWidget( QgsProject *project, QgsMapCanvas *canvas,
...
   mEnableTracingAction->setMenu( tracingMenu );
 
+  // vertex tool may select vertices from all layers
+  mEditVerticesAllLayersAction = new QAction( tr( "Edit vertices of all editable layers at once" ), this );
+  mEditVerticesAllLayersAction->setCheckable( true );
+  mEditVerticesAllLayersAction->setIcon( QIcon( QgsApplication::getThemeIcon( "/mIconEditVerticesAllLayers.svg" ) ) );
+  mEditVerticesAllLayersAction->setToolTip( tr( "The vertex editor tool allows to edit vertices from many layers at once. This option allows to restore previous behavior, ie only edit active layer. Useful with transaction groups and database trigger design interfering with multi layer editing" ) );
+  mEditVerticesAllLayersAction->setObjectName( QStringLiteral( "EditVerticesAllLayersAction" ) );
+  connect( mEditVerticesAllLayersAction, &QAction::toggled, this, &QgsSnappingWidget::enableEditVerticesOnAllLayers );
+
```
Comment (wonder-sk) [inline src/app/qgssnappingwidget.cpp]: I would much rather see this setting somehow connected to vertex tool action (e.g. have a dropdown menu with vertex tool action with this setting, possibly more settings like this in the future)

File: src/app/qgsoptions.cpp [hunk 3]
```diff
@@ -945,6 +945,7 @@ QgsOptions::QgsOptions( QWidget *parent, Qt::WindowFlags fl, const QList<QgsOpti
 
   //default snap mode
   mSnappingEnabledDefault->setChecked( mSettings->value( QStringLiteral( "/qgis/digitizing/default_snap_enabled" ),  false ).toBool() );
+  mEditVerticesOnAllLayersDefault->setChecked( mSettings->value( QStringLiteral( "/qgis/digitizing/default_edit_vertices_on_all_layers_enabled" ), true ).toBool() );
```
Comment (wonder-sk) [inline src/app/qgsoptions.cpp]: I would avoid having this configuration option in global settings dialog...  (even if we store the preferred value in QSettings)

**Comment by pblottiere** [comment]
Thanks to all of you for your comments! 

I added a dropdown menu as @wonder-sk suggested. The PR description has been updated accordingly.

Let me know what you think, and I'll add unit tests as soon as we're agree on the behavior.

**Comment by pblottiere** [comment]
BTW, my icons are not very beautiful and should probably be improved...

File: src/app/vertextool/qgsvertextool.h [hunk 4]
```diff
@@ -61,7 +61,7 @@ class APP_EXPORT QgsVertexTool : public QgsMapToolAdvancedDigitizing
 {
     Q_OBJECT
   public:
-    QgsVertexTool( QgsMapCanvas *canvas, QgsAdvancedDigitizingDockWidget *cadDock );
+    QgsVertexTool( QgsMapCanvas *canvas, QgsAdvancedDigitizingDockWidget *cadDock, bool activeLayerOnly = false );
```
Comment (3nids) [inline src/app/vertextool/qgsvertextool.h]: can you rather create an enum, it's more future proof.
```c++
enum VertexToolMode {
  ActiveLayer,
  AllLayers
};
```

File: src/app/qgisapp.cpp [hunk 5]
```diff
@@ -2828,7 +2831,26 @@ void QgisApp::createToolBars()
...
+  mDigitizeToolBar->insertWidget( mActionDeleteSelected, moveFeatureButton );
+
+  // vertex tool button
+  QToolButton *vertexToolButton = new QToolButton( mDigitizeToolBar );
+  vertexToolButton->setPopupMode( QToolButton::MenuButtonPopup );
+  vertexToolButton->addAction( mActionVertexTool );
+  vertexToolButton->addAction( mActionVertexToolActiveLayer );
+  mDigitizeToolBar->insertWidget( mActionDeleteSelected, vertexToolButton );
+  QAction *defActionVertexTool = mActionVertexTool;
+  switch ( settings.value( QStringLiteral( "UI/defaultVertexTool" ), 0 ).toInt() )
```
Comment (3nids) [inline src/app/qgisapp.cpp]:
```c++
settings.enumValue( QStringLiteral( "UI/defaultVertexTool" ), QgsVertexTool::AllLayers );
```
(or ActiveLayer, but not sure what is agreed as default here, would prefer AllLayers)

File: src/app/qgisapp.cpp [hunk 6]
```diff
@@ -13453,6 +13486,10 @@ void QgisApp::toolButtonActionTriggered( QAction *action )
     settings.setValue( QStringLiteral( "UI/defaultMoveTool" ), 0 );
   else if ( action == mActionMoveFeatureCopy )
     settings.setValue( QStringLiteral( "UI/defaultMoveTool" ), 1 );
+  else if ( action == mActionVertexTool )
+    settings.setValue( QStringLiteral( "UI/defaultVertexTool" ), 0 );
```
Comment (3nids) [inline src/app/qgisapp.cpp]:
```c++
settings.setEnumValue( QStringLiteral( "UI/defaultVertexTool" ), QgsVertexTool::AllLayers );
```

**Review by 3nids — COMMENTED** [review]
some minor things for the settings and API.

but what do we agree as default mode (all vs active layer)?
my vote goes to all.

File: src/app/qgisapp.cpp [hunk 7]
```diff
@@ -2828,7 +2831,26 @@ void QgisApp::createToolBars()
...
+  // vertex tool button
+  QToolButton *vertexToolButton = new QToolButton( mDigitizeToolBar );
+  vertexToolButton->setPopupMode( QToolButton::MenuButtonPopup );
+  vertexToolButton->addAction( mActionVertexTool );
+  vertexToolButton->addAction( mActionVertexToolActiveLayer );
+  mDigitizeToolBar->insertWidget( mActionDeleteSelected, vertexToolButton );
+  QAction *defActionVertexTool = mActionVertexTool;
+  switch ( settings.value( QStringLiteral( "UI/defaultVertexTool" ), 0 ).toInt() )
+  {
+    case 0:
```
Comment (3nids) [inline src/app/qgisapp.cpp]: and of course `case QgsVertexTool::AllLayers:`

File: src/app/vertextool/qgsvertextool.h (hunk 4 above)
Comment (pblottiere) [reply src/app/vertextool/qgsvertextool.h]: You're right, sounds good this way. I'll do that!

File: src/app/qgisapp.cpp (hunk 5 above)
Comment (pblottiere) [reply src/app/qgisapp.cpp]: I don't have a strong opinion here, but I think that we should keep the `AllLayers` mode too.

File: src/app/vertextool/qgsvertextool.h (hunk 4 above)
Comment (pblottiere) [reply src/app/vertextool/qgsvertextool.h]: done

**Comment by pblottiere** [comment]
I took into account previous comments, and Travis is green :).

File: src/ui/qgisapp.ui [hunk 8]
```diff
@@ -1017,7 +1017,10 @@
      <normaloff>:/images/themes/default/mActionVertexTool.svg</normaloff>:/images/themes/default/mActionVertexTool.svg</iconset>
    </property>
    <property name="text">
-    <string>Vertex Tool</string>
+    <string>&amp;Vertex Tool on All Layers</string>
```
Comment (3nids) [inline src/ui/qgisapp.ui]: I think that except the first, the words should be lower case.

**Comment by 3nids** [comment]
LGTM, I am just not 100% sure about capitalisation of the sentence. Any UX expert?

File: src/ui/qgisapp.ui (hunk 8 above)
Comment (pblottiere) [reply src/ui/qgisapp.ui]: I don't have a strong opinion but it seems that all other names are worded in this manner (at least in `src/ui/qgisapp.ui`):

```
...
<string>New 3D Map View</string>
...
<string>New Report</string>
...
<string>Paste Layer/Group</string>
...
```

@3nids What do you think?

Comment (jef-n) [reply src/ui/qgisapp.ui]: @DelazJ What do you think?

Comment (DelazJ) [reply src/ui/qgisapp.ui]: HIG: https://docs.qgis.org/testing/en/docs/developers_guide/hig.html

Comment (nyalldawson) [reply src/ui/qgisapp.ui]: I'd avoid the issue and go "Vertex Tool (All Layers)" and "Vertex Tool (Current Layer)"

**Comment by nyalldawson** [comment]
LGTM - @wonder-sk are you happy with this final approach?

**Comment by wonder-sk** [comment]
Happy :-) feel free to merge...

**Comment by nyalldawson** [comment]
#Super-quick-merge-in-the-hope-that-we-get-below-10-open-PRs

**Comment by pblottiere** [comment]
@wonder-sk @nyalldawson @3nids Thanks guys!

**Pull request merged**